### ISSUE TITLE
Document CameraException error codes.

### ISSUE THREAD

**Issue by amirh** [flutter/flutter#69298]
It's unclear what are the possible values of [CameraException.code](https://github.com/flutter/plugins/blob/8a613b9c55e4b52a7e815eeaafdf39d68176cb72/packages/camera/lib/camera.dart#L131).

**Comment by bharat-biradar** [comment]
HI @pedromassangocode is the explanation of error codes necessary. If it is, can I work on it?

**Comment by pedromassangocode** [comment]
Hi @bharat-biradar 
Sorry, I'm not sure. Maybe @amirh could give you a hint.

**Comment by bharat-biradar** [comment]
@amirh is the explanation required, if not this issue can be closed.

**Comment by huberrom** [comment]
Hi  @bharat-biradar @pedromassango ! 

I'm on this issue because the error codes was necessary, and I indeed didn't know which code was equal to permission not granted.

For me it is needed if the user refuse to give permission (and select the option "Don't ask again"). I want to react to this event saying my app need it and then open settings.

Maybe a enum could be created so we can check which error we have (and to prevent futur bug if you change the error code).

Thanks in advance !

**Comment by bharat-biradar** [comment]
@pedromassangocode Shall I document the error codes and open a PR.

**Comment by pedromassangocode** [comment]
@bharat-biradar sure go-ahead :)

**Comment by bharat-biradar** [comment]
Sorry for the delay, I'll add a PR by the day's end

**Comment by bharat-biradar** [comment]
@huberrom the error code `"cameraPermission"` may be due given due to different reasons.

I found a lot of error codes and their description. Just to confirm is this what you needed to be documented.

```
///  |Error code          |  Error Description    |
  ///  "cameraPermission" => "Camera permission request ongoing"
  ///  "cameraPermission" => "MediaRecorderCamera permission not granted"
  ///  "cameraNotFound"   => "Camera not found. Please call the 'create' method before calling 'initialize'."
  ///  "captureAlreadyActive" => "Picture is currently already being captured"
  ///  "cannotCreateFile"  => "failed to create a file" (may depend on various factors)
  ///  "captureTimeout" => "Picture capture request timed out"
  ///  "captureFailure" => "Unknown reason"
  ///  "captureFailure" => "The capture has failed due to an abortCaptures() call"
  ///  "captureFailure" => "An error happened in the framework"
  ///  "IOError"     => "Failed saving image"
  ///  "IOError"  => "Setup Writer Failed"
  ///  "videoRecordingFailed"  => "Failed to record video"  (may depend on various factors)
  ///  "videoRecordingFailed"  => "pauseVideoRecording requires Android API +24."
  ///  "videoRecordingFailed"  => "resumeVideoRecording requires Android API +24."
  ///  "setFlashModeFailed"  => "Device does not have flash capabilities"
  ///  "setFlashModeFailed" => "Unknown flash mode " + mode
  ///  "setFlashModeFailed"   => "Could not set flash mode."
  ///  "setExposureModeFailed" => "Unknown exposure mode " + mode
  ///  "setExposurePointFailed" => "Device does not have exposure point capabilities"
  ///  "setExposurePointFailed" => "Could not determine max region boundaries"
  ///  "setFocusModeFailed" => "Unknown focus mode " + mode
  ///  "setFocusPointFailed" => "Device does not have focus point capabilities"
  ///  "setFocusPointFailed" => "Could not determine max region boundaries"
  ///  "ZOOM_ERROR"  =>   "Zoom level out of bounds (zoom level should be between min_zoom and max_zoom)."
  ///  "ZOOM_ERROR"  =>  "setZoomLevel is called without specifying a zoom level."
```

**Comment by huberrom** [comment]
Yes this is exactly what I need (in fact I personnaly just need cameraPermission).

I got another error code which is "Error -11852", which I got only on IOS and append at the same time as cameraPermission on Android. I don't where it come from tho

**Comment by mariopepe** [comment]
If I disable in the settings the permissions for camera then I get either one of these two errors when I try to initialise the camera:

```
cameraException.code = Error -11852
cameraException.description: Cannot use Front Camera
```

or 

```
cameraException.code = Error -11852
cameraException.description: Cannot use Back Camera
```

Which one I get depends on which cameraLensDirection I try to initialise.

Of course since I was purposely testing this, now I will write some code to handle `Error -11852` so to ask my users to re-enable the permission, but there seems to be two issues here:
1.  It seems that the plugin does not recognise that the underlying problem are the missing permissions, the description Cannot use Front Camera could be considered misleading or at least not very on point
2. I tried my best but I couldn't find the documentation of the CameraException error codes (the topic of this issue)

Informing the user of the state of the application is clearly very important, especially during failure states.
In order to achieve a clear representation of failure states, precisely mapped errors are very important in general.
But in this specific case (i.e. the camera missing permission), it is of the utmost importance to capture this error properly as it is something the user could fix right away if prompted correctly from the developer.

Having said that, I wanted to ask if it was possible to bump up the P5 priority for this, and if not, if someone would be so kind to explain me why I am wrong and why a low prioritisation is better in this case! If you have no time feel free to send me the link for the Flutter Team documentation regarding prioritisation levels, so I can understand it by myself, I searched for it but had no luck.

Thanks so much for your help!

**Comment by gmcdowell** [comment]
Any update on when this might get documented and code enum setup?

**Comment by alestiago** [comment]
I really do not like the magic string error codes. Could we explore the possibilities on making this an enumeration so IDE autocompletion can document it. The advantages I can see are:

- Each enumeration value can have dart doc to specify the error.
- Maintainability.
- IDE support.
- Self documenting.

Another option would be subclassing CameraException. Advantages I can see here:
- Catching specific exceptions is made easier within try catch
- Once can check subclasses of CameraException to see what are the possible Exceptions
- Document each method with what subclasses of CameraException it may throw

**Comment by akbarnotopb** [comment]
> @huberrom the error code `"cameraPermission"` may be due given due to different reasons.
> 
> I found a lot of error codes and their description. Just to confirm is this what you needed to be documented.
> 
> ```
> ///  |Error code          |  Error Description    |
>   ///  "cameraPermission" => "Camera permission request ongoing"
>   ///  "cameraPermission" => "MediaRecorderCamera permission not granted"
>   ///  "cameraNotFound"   => "Camera not found. Please call the 'create' method before calling 'initialize'."
>   ///  "captureAlreadyActive" => "Picture is currently already being captured"
>   ///  "cannotCreateFile"  => "failed to create a file" (may depend on various factors)
>   ///  "captureTimeout" => "Picture capture request timed out"
>   ///  "captureFailure" => "Unknown reason"
>   ///  "captureFailure" => "The capture has failed due to an abortCaptures() call"
>   ///  "captureFailure" => "An error happened in the framework"
>   ///  "IOError"     => "Failed saving image"
>   ///  "IOError"  => "Setup Writer Failed"
>   ///  "videoRecordingFailed"  => "Failed to record video"  (may depend on various factors)
>   ///  "videoRecordingFailed"  => "pauseVideoRecording requires Android API +24."
>   ///  "videoRecordingFailed"  => "resumeVideoRecording requires Android API +24."
>   ///  "setFlashModeFailed"  => "Device does not have flash capabilities"
>   ///  "setFlashModeFailed" => "Unknown flash mode " + mode
>   ///  "setFlashModeFailed"   => "Could not set flash mode."
>   ///  "setExposureModeFailed" => "Unknown exposure mode " + mode
>   ///  "setExposurePointFailed" => "Device does not have exposure point capabilities"
>   ///  "setExposurePointFailed" => "Could not determine max region boundaries"
>   ///  "setFocusModeFailed" => "Unknown focus mode " + mode
>   ///  "setFocusPointFailed" => "Device does not have focus point capabilities"
>   ///  "setFocusPointFailed" => "Could not determine max region boundaries"
>   ///  "ZOOM_ERROR"  =>   "Zoom level out of bounds (zoom level should be between min_zoom and max_zoom)."
>   ///  "ZOOM_ERROR"  =>  "setZoomLevel is called without specifying a zoom level."
> ```

Still no documentation about this. as per `10.1` access denied code was CameraAccessDenied

**Comment by gmackall** [comment]
@camsim99 just wanted to cc you on this as it will be relevant when we end up making changes to how/when we error in the camerax plugin
